finetune/lora.py

```diff
@@ -0,0 +1,225 @@
+    #    # strict=False because missing keys due to adapter weights not contained in state dict
+    #    model.load_state_dict(checkpoint, strict=False)
+
+    with fabric.init_module(), lora(r=lora_r, alpha=lora_alpha, dropout=lora_dropout, enabled=True):
+        model = Parrot(config)
+        # strict=False because missing keys due to LoRA weights not contained in checkpoint state
+        model.load_state_dict(torch.load(checkpoint_path), strict=False)
+
+    mark_only_lora_as_trainable(model)
+    num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
+    fabric.print(f"Number of trainable parameters: {num_params}")
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
+    model, optimizer = fabric.setup(model, optimizer)
+    train(fabric, model, optimizer, train_data, val_data, checkpoint_dir, out_dir)
+
+    # Save the final LoRA checkpoint at the end of training
+    checkpoint = lora_state_dict(model)
+
+    fabric.save(os.path.join(out_dir, "lit-model-lora-finetuned.pth"), checkpoint)
+
+
+def train(
+    fabric: L.Fabric,
+    model: torch.nn.Module,
+    optimizer: torch.optim.Optimizer,
+    train_data: np.ndarray,
+    val_data: np.ndarray,
+    checkpoint_dir: Path,
+    out_dir: str,
+) -> None:
+    """The training loop.
+
+    Loosely based on the nanoGPT implementation: https://github.com/karpathy/nanoGPT.
+    """
+
+    step_count = 0
+
+    tokenizer = Tokenizer(checkpoint_dir / "tokenizer.json", checkpoint_dir / "tokenizer_config.json")
+
+    for iter_num in range(max_iters):
+
+        if step_count <= warmup_iters:
+            # linear warmup
+            lr = learning_rate * step_count / warmup_iters
+            for param_group in optimizer.param_groups:
+                param_group['lr'] = lr
+
+        t0 = time.time()
+
+        input_ids, targets = get_batch(fabric, train_data)
+        with fabric.no_backward_sync(model, enabled=((iter_num + 1) % gradient_accumulation_iters != 0)):
+            logits = model(input_ids)
+            loss = loss_fn(logits, targets)
+            fabric.backward(loss / gradient_accumulation_iters)
+
+        if (iter_num + 1) % gradient_accumulation_iters == 0:
+            optimizer.step()
+            optimizer.zero_grad()
+            step_count += 1
+
+            if step_count % eval_interval == 0:
+                val_loss = validate(fabric, model, val_data, tokenizer)
+                fabric.print(f"step {iter_num}: val loss {val_loss:.4f}")
+                fabric.barrier()
+
+            if step_count % save_interval == 0:
+                print(f"Saving LoRA weights to {out_dir}")
+                # We are only saving the LoRA weights
+                # TODO: Provide a function/script to merge the LoRA weights with pretrained weights
+                checkpoint = lora_state_dict(model)
+                fabric.save(os.path.join(out_dir, f"iter-{iter_num:06d}-ckpt.pth"), checkpoint)
```

Comment:
This isn't how you would use `fabric.save`. See the docs: https://lightning.ai/docs/fabric/latest/api/fabric_methods.html#save

You probably want to use `torch.save` instead for now. We'll be working on a way to filter out what gets saved soon. Then you'll be able to use `fabric.save`